Add ethereum beacon light client

ParthDesai opened the pull request:
## Description
This PR adds ethereum beacon light client pallet and its primitive to the repository. The pallet is not part of any runtime at the moment. Future PRs will create a new core domain runtime and add it.

The pallet is heavily modified version of snowfork's ethereum beacon client with several improvements on top of it.

## How it works? 
There are three main extrinsics:

### light_client_update
As name suggests light client update extrinsic updates the state of light client. Specifically, if light client update is valid, the sync committees and/or finalized header will be updated. 

### import execution header
Import execution header verifies if the incoming execution header is valid against the stored sync committees and if it is valid it is stored as execution header. This can later be used to verify contract storage.

### unblock bridge
Unblock bridge extrinsic can only be called by root. In production, this should only happen via governance.  This extrinsic is incomplete at the moment and only set the bridge blocked status to false. It needs to be expanded in future to ingest governance approved light client update.


## Configuration
Two disjoint set of configurations:

### Pallet configuration
Some configuration like Fork versions, prune threshold are configured at the pallet's config trait. 

### Const parameter in codebase
Configuration that can very across network and is not going to change for one network is defined as constant parameter in codebase. 

### Why two different configuration
Configuration pertaining to the size of the data structure inside the sync data needs to be a constant and we cannot use generic parameter inside the expression. The restriction comes from the library used to serialize and deserialize data structure to/from simple serialization format. (Unless we enable generic_const_exprs unstable feature, which is not ideal).

## What we can improve upon?
- The library `ssz` used here is not production grade. It is better to use `lighthouse`'s ssz library. Added benefit of using this library is we can consolidate the configuration to pallet config as well. 

## Reference to the specification
https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md

Closes #1202

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

vedhavyas reviewed (commented):
I did not look at all the changes. Just did a brief pass the PR. This is really tough for me to understand the changes. I'm not sure how this is different from the Snowfork's. I would ideally preferred just the plain pallet logic ignoring benchmarking and other testdata as these could have come in later PRs. 

Is it possible to break the commit in smaller commits so I understand the implementation from the spec that is on the Notion ? Ideally it these changes could come in multiple PRs

ParthDesai committed:
snowbridge ethereum beacon client pallet + primitives
project:github.com/snowfork/snowbridge commit:2ed11558c09d7e061c311244c4922c98ad915c50

ParthDesai committed:
fix code formatting to minimize diff in subsequent commits

ParthDesai committed:
Changes into helper function defined in pallet implementation
This includes addition of light client update validation + supporting functions for that

ParthDesai committed:
Changes into pallet extrinsics
Removal of finalized header extrinsic (the functionality is merged into light client update now)

ParthDesai committed:
Have unified network parameter configuration
As opposed to two set of configuration: One within the pallet config and other as constants

ParthDesai committed:
changes in pallet errors
More granual errors are defined and unused errors are cleaned up.

ParthDesai committed:
Modify storage to support pruning of Sync committees and execution headers

ParthDesai committed:
Refactoring: Add sync committee update event
Add documentation comment for errors and pallet storage
Tidying up the pallet Renaming variables + Missing imports + Clippy fixes

ParthDesai committed:
Fixing tests as per pallet changes and adding mainnet tests

ParthDesai committed:
light client data structure changes

ParthDesai committed:
benchmarking code + new weights

ParthDesai committed:
Tidying up project
Updating cargo files + removing not required file + Applying feedback

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -341,139 +341,291 @@ pub mod pallet {
+            update.finality_branch.len() != 0
+        }
+
+        fn validate_light_client_update(
```
Can you explain what light client is here? This seems to be introduced in this commit

ParthDesai replied:
https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#lightclientupdate

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -216,84 +216,50 @@ pub mod pallet {
 
-        #[pallet::weight(T::WeightInfo::import_finalized_header())]
-        #[transactional]
-        pub fn import_finalized_header(
```
Do we not need this anymore ?

ParthDesai replied:
This is not supposed to be a separate endpoint as per the ethereum specification. The `light client update` endpoint updates the entire state of light client (except execution header import).

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -216,84 +216,50 @@ pub mod pallet {
-
+        /// Import an execution header
         #[pallet::weight(T::WeightInfo::import_execution_header())]
         #[transactional]
```
this is defualt. We dont need to have it explicitly IMO

ParthDesai replied:
Make sense. will remove.

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -216,84 +216,50 @@ pub mod pallet {
                 log::error!(
                     target: "ethereum-beacon-client",
-                    "💫 Sync committee period update failed with error {:?}",
+                    "💫 Light client update failed with error {:?}",
```
Not sure I understand this either. Why the refactor ?

ParthDesai replied:
[See my comment](https://github.com/subspace/subspace/pull/1170#issuecomment-1453590351)

nazar-pc replied:
Did you suggest these renamings to Slowfork?

vedhavyas commented on domains/primitives/beacon/src/lib.rs:
```diff
@@ -174,7 +174,7 @@ pub struct InitialSync<SyncCommitteeSize: Get<u32>, ProofSize: Get<u32>> {
 #[scale_info(skip_type_params(SignatureSize, ProofSize, SyncCommitteeSize, SyncCommitteeSize))]
 #[codec(mel_bound())]
-pub struct SyncCommitteePeriodUpdate<
+pub struct LightClientUpdate<
```
Why the change here ?

ParthDesai replied:
https://github.com/subspace/subspace/pull/1170#issuecomment-1453590351

vedhavyas reviewed (commented):
Had a quick look over. Great work 👍🏼 .I have ignored first commit. Just did a brief check as the changes are far too big and I do have little understanding of the snowfork's pallet. 

Some of the commits could be merged since either they are too small or scattered in between other commits. I was expecting the following order of commits for this PR
1 - Snowfork fork
2 - general pallet documentation changes and fmt, warnings as per our project rules. 
... rest of the commits following our contribution guidelines.

I have a few questions on `LightClient` you are referring to? I'm not sure if calling it light client would be right word though. I think Snowfork terminology was sufficiently closer to the actual ethereum terminology and easier for readability.

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -341,139 +341,291 @@ pub mod pallet {
+                    && (Self::compute_sync_committee_period(update.finalized_header.slot)
+                        == Self::compute_sync_committee_period(update.attested_header.slot));
+
+            if (Self::get_sync_committee_sum(sync_committee_bits.to_vec()) * 3
```
So the threshold (2/3) is a constant in the ethereum spec? A comment or a smaller helper function could be likely better than directly using the magic number.

ParthDesai replied:
Yes. It is constant in the spec.

ParthDesai replied:
I have left comment there. And for efficiency purpose removed the check as it was done by validate light client as well.

liuchengxu reviewed (commented):
Congrats the achivements here! Just a few general comments as I'm not familar with this part.

- Looks like the most substantial changes are introduced in the third and fourth commit, but their commit messages are unfortunatley not informative as the later commits :P. It would be very helpful for me who has little knowledge on this front if they can splitted into several smaller featureful commits with a detailed commit message to convey what the changes are for.
- Nicer to move the fifth commit before the third one so that we won't see the code in config from nowhere when looking at the third commit.
- Several minor refactoring commits later can be merged into one with a commit message that says no substantial changes, just various minor refactorings, so that we can get through it pretty quickly.

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -344,139 +344,291 @@ pub mod pallet {
+
+            if !Self::is_sync_committee_update(update) {
+                ensure!(
+                    update.next_sync_committee == Default::default(),
```
Could be more readable for me if it's something like `ensure!(update.next_sync_committee.is_empty(), Error::<T>::NonEmptySyncCommittee)`, which is more consistent with the error type as well. There are also a few other similar places, it's kind of weird to check `foo == Default::default()` IMHO.

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1093 @@
+
+        /// A period in which a the light client update must be submitted to this light client.
+        /// In case we did not receive it, the light client will be blocked.
+        type WeakSubjectivityPeriodSeconds: Get<u32>;
```
Missing `#[pallet::constant]`?

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1093 @@
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
```
Perhaps simply remove it if unused.

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1093 @@
+    impl<T: Config> GenesisBuild<T> for GenesisConfig {
+        fn build(&self) {
+            log::info!(
+                target: "ethereum-beacon-client",
```
We typically add a `runtime` prefix for the target in pallets `runtime::ethereum-beacon-client`.

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1093 @@
+            fork_versions.genesis.version
+        }
+
+        pub(super) fn initial_sync(initial_sync: InitialSyncOf) -> Result<(), &'static str> {
```
`initial_sync` and `process_initial_sync` seem to be std only.

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1032 @@
+		) -> DispatchResult {
+			let sync_committee_sum = Self::get_sync_committee_sum(sync_committee_bits.clone());
+			ensure!(
+				(sync_committee_sum * 3 >= sync_committee_bits.clone().len() as u64 * 2),
```
I would make these magic values constants

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/config.rs:
```diff
@@ -0,0 +1,39 @@
+pub const MAX_ATTESTER_SLASHINGS: usize = 2;
+pub const MAX_ATTESTATIONS: usize = 128;
+pub const MAX_DEPOSITS: usize = 16;
+pub const MAX_VOLUNTARY_EXITS: usize = 16;
```
would be great to add some comments explaining some of the constants here that are not self explanatory

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1032 @@
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::sync_committee_period_update())]
+		#[transactional]
+		pub fn sync_committee_period_update(
```
can you add more comments for the these Calls as well ?

ParthDesai replied:
Comments are there in the newer commit (the commit which adds comments to errors and calls and such). I separated out that commit  otherwise commit that is making the changes into this file would be too big.

vedhavyas commented on domains/pallets/ethereum-beacon-client/src/ssz.rs:
```diff
@@ -0,0 +1,155 @@
+use ssz_rs_derive::SimpleSerialize;
+
+#[derive(Default, Debug, SimpleSerialize, Clone)]
+pub struct SSZVoluntaryExit {
```
add more comments to these structures pls

ParthDesai replied:
Added comments to the original generic data structures.

vedhavyas commented on domains/primitives/beacon/Cargo.toml:
```diff
@@ -0,0 +1,37 @@
+[package]
```
i think the crate name should be more than just `beacon`. Just beacon is not clear what is it for

ParthDesai replied:
The crate's name is `snowbridge-beacon-primitives`

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1093 @@
+            );
+
+            let stored_latest_finalized_header_state = <LatestFinalizedHeaderState<T>>::get();
+            let store_period = Self::compute_sync_committee_period(
```
Could somthing like `last_finalized_period` be more descriptive?

ParthDesai replied:
The name was chosen to be in parity with the spec code. See here: [spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#validate_light_client_update)

vedhavyas reviewed (commented):
Minor nitpicks. Make sense overall

liuchengxu commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -0,0 +1,1093 @@
+            origin: OriginFor<T>,
+            light_client_update: LightClientUpdateOf,
+        ) -> DispatchResult {
+            let _sender = ensure_signed(origin)?;
```
I guess this and `import_execution_header` below will be changed to the relayer only in the future?

vedhavyas replied:
is that required to be relayer though? Not sure if we need to maintain relayer set

liuchengxu replied:
I recall somewhere in the design doc the relayer was mentioned, thus had this question, I could be wrong.

ParthDesai replied:
@vedhavyas @liuchengxu This will depend upon whether we want relayer incentivization in near future or not. My personal opinion is that we do not need to implement it now and anybody who can pay the tx fee can submit the *untrusted* data to the light client.

liuchengxu reviewed (commented):
Reasonable in generall, just several minor comments and questions.

ParthDesai committed:
address minor nitpicks + better readability

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -1,32 +1,31 @@
 //! # Ethereum Beacon Client
 #![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::large_enum_variant)] // Runtime-generated enums
```
Can you suppress it on a specific data structure rather than the whole crate? Ideally fixing it, of course.

ParthDesai replied:
This is happening due to three variants of `pallet::call` enum generated by the macro differs in size. Reason being that the bridge blocking call do not take any argument while other two call's arguments are pretty big. 

Since the enum is generated by runtime, I do not think I can suppress it for that specific structure.

nazar-pc replied:
But you can box large enum variant contents to make it not be as large anymore. We have it in a few pallets, you can just search for `Box` in them and use as an example.

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -32,36 +32,46 @@ use sp_std::prelude::*;
+    config::MaxAttestations,
+    config::MaxValidatorsPerCommittee,
+    config::SyncCommitteeSize,
 >;
```
This feels like a downgrade. The reason it was done as config constants was such that it can be customized in the runtime. Now you have hardcoded the values. Moreover, you made selection depend on feature flag, which with cargo unification features means you can't have this crate used in one runtime configured for mainnet and another for testnet, it will force both to use one of those two, which is clearly suboptimal and we'll have to revert these changes sooner or later.

ParthDesai replied:
Actually, snowfork had two exact same set of configuration. One as constant and other configurable via pallet config. It was confusing as if you change the constants but do not change the pallet config the pallet won't work. 

Configuration pertaining to the size of the data structure inside the sync data needs to be a constant and we cannot use generic parameter inside the expression. The restriction comes from the library used to serialize and deserialize data structure to/from simple serialization format. (Unless we enable generic_const_exprs unstable feature, which is not ideal).

So, I just removed the configuration from the runtime and used constant parameter types to have same set of configuration. Yes, this restricts our ability to use the pallet, this is something I need to fix later on.

nazar-pc replied:
'generic_const_exprs' is not an issue, we use a few unstable features already. If there was duplication, I'd imagine it to be de-duplicated in a way that extends functionality rather than hardcoding it and would upstream that to Snowfolk.

For instance, change constants to runtime configuration, that way all of those parameters could be configured in the chain spec, not requiring any features on pallet level and being as flexible as we want. But at least having runtime constants configurable from the outside is not as bad as just hardcoding them and making it impossible to have two runtimes in the workspace that require different feature flags.

ParthDesai replied:
Hmm. But aren't unstable feature forward incompatible? It may very well possible in future that `generic_const_expr`'s behavior is altered or it is removed altogether.

nazar-pc replied:
If that ever happens we'll switch to an alternative. It is unlikely though for many features, this one including.

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -193,12 +203,6 @@ pub mod pallet {
     }
 
     #[cfg(feature = "std")]
```
This feature was applied to `the impl, if you decided to remove this (though I genuinely don't understand why, it seems helpful to be able to configure in chain spec) it needs to go away too. I guess `#[pallet::genesis_build]` adds the same attribute too, which is why it still works by accident (or does it? maybe we just don't use it a thus don't know).

ParthDesai committed:
Merge branch 'main' into integrations/add-ethereum-beacon-light-client

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -156,10 +164,7 @@ pub mod pallet {
 
-    #[pallet::storage]
-    pub(super) type FinalizedBeaconHeaders<T: Config> =
-        StorageMap<_, Identity, H256, BeaconHeader, OptionQuery>;
```
Why was it deleted, is it not used?

ParthDesai replied:
Yes

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -660,18 +805,15 @@ pub mod pallet {
-            sync_committee: SyncCommitteeOf<T>,
-            sync_committee_branch: BoundedVec<H256, T::MaxProofBranchSize>,
+            sync_committee_root: H256,
+            sync_committee_branch: BoundedVec<H256, config::MaxProofBranchSize>,
```
This and some other changes seem to belong to the previous commit. My guess is that previous commit doesn't compile and gives incomplete impression of what refactoring contained.

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -944,105 +1105,5 @@ pub mod pallet {
-        }
-    }
-
-    impl<T: Config> Verifier for Pallet<T> {
```
Why was this implementation removed? I don't see anything that removes its usage, imports, it seems to be unrelated to what is changing here.

ParthDesai replied:
Verifier is used by snowfork to expose interface that allows the message receiver to verify if message is indeed received from the ethereum blockchain by verifying the storage proof. I tested their implementation which is broken atm and beside our implementation will be more likely different.

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -155,8 +152,26 @@ pub mod pallet {
-        ExecutionHeaderNotLatest,
+        /// Bridge is blocked at the moment
         BridgeBlocked,
+        /// Slot ordering of light client update is invalid
+        InvalidSlotsOrdering,
+        /// Sync committee signed the attested header is not recognized
+        UnrecognizedSyncCommittee,
+        /// Light client update does not contain any new information
+        NonRelevantUpdate,
+        /// Light client update does not contain enough information to be considered
+        NotApplicableUpdate,
+        /// Light client update should not contain finalized header
+        NonEmptyFinalizedHeader,
+        /// Light client update should not contain sync committee
+        NonEmptySyncCommittee,
+        /// Next sync committee in light client does not match stored committee for that period
+        NextSyncCommitteeMismatch,
+        /// Finalized header's sync committee period does not match store's current period
+        FinalizedPeriodMismatch,
+        /// Unable to store non sequential sync committee update
+        NonSequentialSyncCommitteeUpdate,
```
Where are these actually used? This commit just introduces a bunch of variants out of nowhere.

nazar-pc commented on domains/pallets/ethereum-beacon-client/src/lib.rs:
```diff
@@ -180,13 +180,17 @@ pub mod pallet {
     #[pallet::storage]
     pub(super) type ExecutionHeaders<T: Config> =
-        StorageMap<_, Identity, H256, ExecutionHeaderOf<T>, OptionQuery>;
+        CountedStorageMap<_, Identity, H256, ExecutionHeaderOf, OptionQuery>;
```
Commit says that this is modified, I see that, but why?

nazar-pc reviewed (changes requested):
One this is that I don't understand the Ethereum light client itself, so I can've verify business logic, but even on superficial level these changes do not make complete sense. I had an impression that it shouldn't compile and indeed when cloned locally even dependencies are impossible to install until commit "Tidying up project
Updating cargo files + removing not required file + Applying feedback". I'm really not sure how you were developing this all the way through, but I'd expect the very first commit after importing unmodified code of Snowfolk to make it at least pass `cargo check` and hopefully clippy/formatting shortly afterwards. Without that it was possible to introduce a bunch of changes that were supposed to be atomic, but were in practice spread out across multiple commits, so as a reviewer I'm confused and it doesn't make a lot of sense.

https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md was written to be used as a guidance, especially in such large refactoring projects like this one. If Snowfolk compiled and worked upstream, our fork should also compile and work from second commit, not from 12th.

Imagine we have an issue with these changes somewhere and we need to go back and do `git bisect` to find the point where things went wrong. We wouldn't be able to because we can't even install dependencies, let alone compile it.
